File: javascript/data_structures/queue/queue.js
class Queue {
    constructor() {
        this.first = null;
        this.last = null;
    }

    enqueue(data) {
        const newNode = {
            data: data,
            next: null
        }

        if(!this.first) {
            this.first = newNode;
            this.last = newNode;
            return;
        }

        this.last.next = newNode;
        this.last = this.last.next;
    }

    dequeue() {
        if(!this.last)
            throw new Error("Cannot dequeue empty queue");
        
        const returnValue = this.first.data;
        this.first = this.first.next;

        if(this.first === null)
            this.last = null;
        return returnValue;
    }

    isEmpty() {
        return Boolean(!this.first);
    }

    peek() {
        if(!this.first)
            throw new Error("Cannot peek empty queue");
        return this.first.data;
    }
}

module.exports = Queue;

let queue = new Queue();

console.assert(queue.isEmpty() === true);
queue.enqueue(1);
console.assert(queue.peek() === 1);
console.assert(queue.isEmpty() === false);
queue.enqueue(2);
console.assert(queue.peek() === 1);
queue.enqueue(3);
console.assert(queue.dequeue() === 1);
console.assert(queue.peek() === 2);
console.assert(queue.dequeue() === 2);
console.assert(queue.peek() === 3);
console.assert(queue.isEmpty() === false);
queue.enqueue(4);
queue.enqueue(5);
console.assert(queue.peek() === 3);
console.assert(queue.dequeue() === 3);
console.assert(queue.peek() === 4);
console.assert(queue.dequeue() === 4);
console.assert(queue.peek() === 5);
console.assert(queue.dequeue() === 5);
console.assert(queue.isEmpty() === true);

File: python/data_structures/queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, value):
        if self.head is None:
            self.head = Node(value)
            self.tail = self.head
        else:
            self.tail.next = Node(value)
            self.tail = self.tail.next

    def dequeue(self):
        if self.head is None:
            return
        return_value = self.head.value
        self.head = self.head.next
        if self.head is None:
            self.tail = None
        return return_value


queue = Queue()
queue.insert(1)
queue.insert(2)
queue.insert(3)
queue.insert(4)

assert queue.dequeue() == 1
assert queue.dequeue() == 2
queue.insert(5)
assert queue.dequeue() == 3
assert queue.dequeue() == 4
assert queue.dequeue() == 5
